### PR TITLE
[Extensions] Add Regex validator

### PR DESCRIPTION
**Pull request by mloufra**
### Description
Add Regex Validator implement Writeable and Validator to make Validator writeable for custom setting. 

### Issues Resolved
Fixes https://github.com/opensearch-project/opensearch-sdk-java/issues/350

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/common/settings/Setting.java:1315]: This is good.  Compilation is the most performance-sensitive part so it is much better to store the compiled Pattern here rather than the string.

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:219]: Great job with the logic here.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Great job!   

A few minor tweaks needed, and a usability suggestion from the original issue.

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (mloufra) [reply server/src/main/java/org/opensearch/common/settings/Setting.java:1312]: ACK

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (mloufra) [reply server/src/main/java/org/opensearch/common/settings/Setting.java:1309]: ACK

File: server/src/main/java/org/opensearch/common/settings/WriteableSetting.java
Comment (mloufra) [reply server/src/main/java/org/opensearch/common/settings/WriteableSetting.java:99]: ACK

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/common/settings/Setting.java:1329]: These two `throw` statements don't seem to make sense.  If it's not null (the usual case) they just see that it's not valid, but aren't told why.  You'd only get to the second `throw` if the value was null, in which case it would not match the regex.

I think you just need a single `throw` statement (the second one).  If you want to add the value to the string you can do that to combine the two strings, but make only one throw.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:11]: This slipped in?

**Review by dblock — CHANGES_REQUESTED** [review]
This code needs tests, otherwise LGTM.

File: CHANGELOG.md
Comment (mloufra) [reply CHANGELOG.md:11]: It should comes from rebase mistake, I will delete that in next commit.

**Review by owaiskazi19 — APPROVED** [review]

**Review by ryanbogan — APPROVED** [review]

File: server/src/test/java/org/opensearch/common/settings/SettingTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/common/settings/SettingTests.java:337]: Use `expectedPattern.pattern()` rather than the `toString()`.  They may be the same thing, but the `toString()` is intended to be human-readable while `pattern()` explicitly returns the stored string.

File: server/src/test/java/org/opensearch/common/settings/SettingTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/common/settings/SettingTests.java:338]: You need to add a test of the validator itself.  

Since you've given it a test pattern of at least one digit, any string with just digits should throw an exception and any other string should not.

```java
assertThrows(IllegalArgumentException.class, () -> regexValidator.validate("123"));
assertDoesNotThrow(() -> regexValidator.validate("abc"));
```

Note the pattern you give is what fails the validation so if you want to ensure it's just digits you could use the `\\D+` pattern to fail if it's anything other than digits and reverse the above two tests.

Repeat these tests below after writing and reading the parser.

**Review by dbwiddis — COMMENTED** [review]
Need to test the validator throwing the exception.

File: server/src/test/java/org/opensearch/common/settings/SettingTests.java
Comment (mloufra) [reply server/src/test/java/org/opensearch/common/settings/SettingTests.java:344]: Try to Use `try {
    regexValidator.validate("abc");
} catch (IllegalArgumentException e) {
    fail("Expected validate() to not throw an exception, but it threw " + e);
}
` instead of 
`assertThrows(IllegalArgumentException.class, () -> regexValidator.validate("123"));
assertDoesNotThrow(() -> regexValidator.validate("abc"));` 
because the error message shows 
`> Task :client:test:loggerUsageCheck
        assertDoesNotThrow(() -> regexValidator.validate("abc"));
        ^
  symbol:   method assertDoesNotThrow(()->regexV[...]abc"))
  location: class SettingTests
/home/runner/work/OpenSearch/OpenSearch/server/src/test/java/org/opensearch/common/settings/SettingTests.java:348: error: cannot find symbol
                assertDoesNotThrow(() -> regexValidator.validate("abc"));
                ^
  symbol:   method assertDoesNotThrow(()->regexV[...]abc"))
  location: class SettingTests`
after doing some research I find out that `assertDoesNotThrow()` is a JUnit 5 feature that was introduced in version 5.2.0, so if we are using an earlier version of JUnit 5, the method will not be available. But I am not able to find how to check the Junit version, so I wrap the call to validate() in a try-catch block and use fail() to signal that an exception was thrown.

Comment (dbwiddis) [reply server/src/test/java/org/opensearch/common/settings/SettingTests.java:342]: Ah, correct. I knew they added `assertThrows()` in JUnit 4.13 (which is used here) but didn't know the opposite didn't work.  I think your try/catch and fail in the catch block is a good option instead.

Comment (dbwiddis) [reply server/src/test/java/org/opensearch/common/settings/SettingTests.java:342]: @mloufra you added the "not throws" case but didn't add the assertion for the expected exception.

Comment (mloufra) [reply server/src/test/java/org/opensearch/common/settings/SettingTests.java:342]: ACK

File: server/src/test/java/org/opensearch/common/settings/SettingTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/common/settings/SettingTests.java:345]: This test is [failing Gradle Check](https://build.ci.opensearch.org/job/gradle-check/11913/testReport/junit/org.opensearch.common.settings/SettingTests/testRegexValidator/):
> java.lang.AssertionError: Expected validate() to not throw an exception, but it threw java.lang.IllegalArgumentException: Setting must be valid [\d+],Custom setting matches regex [\d+]

You're trying to send in a string (which happens to be a regex but isn't treated as one there) as the setting value and it's failing (as expected).

Using a regular expression as the string to check is very confusing to other maintainers and does not correspond with the common use case.

The only place you should use the regex is constructing the validator, which you do with `new RegexValidator(expectedRegex)`.  Combined with the check `if (!pattern.matcher(value).matches())` that means you will fail if the regex does *not* match the input.  

The `\d+` regex matches 1 or more digits.  (In Java we need the double slashes to escape it for the string `"\\d+".)

So you should pass for something that is 1 or more digits like "123".  Just use "123" on line 343.  And use some other text string like "foo" or "text" on line 340 above.

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/common/settings/Setting.java:1259]: Add a javadoc for this constructor to tell users what the string means.  "A regular expression containing the only valid input for this setting."

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/common/settings/Setting.java:1275]: This message doesn't match the failure case. I think I suggested this in an issue or PR comment somewhere else when I thought you were failing if it matched the regex.  With the `!` you are failing if it does NOT match.  That's fine but this error message should be `"Setting [" + value + "] does not match regex [" + pattern.pattern() + "]"`.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/common/settings/Setting.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/common/settings/Setting.java:1259]: You added the javadoc to the streaminput constructor below, not this one... this is the one developers will use.  The StreamInput constructor will be used by the transport service sending these between OpenSearch and SDK.

Also you need the `@param` javadoc tag for the parameters.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Some final javadoc fixes and this will be GTG

**Review by dbwiddis — APPROVED** [review]

**Pull request merged**
